File: 543.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    int ans = 0;
    public int diameter(TreeNode root){
        if(root==null){
            return 0;
        }
        int lh = diameter(root.left);
        int rh = diameter(root.right);

        ans = Math.max((lh+rh), ans);
        return Math.max(lh,rh)+1;
    }
    public int diameterOfBinaryTree(TreeNode root) {
        diameter(root);
        return ans;
    }
}

File: 543.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int fDiam(TreeNode* root, int& maxi){
        if(root == NULL){
            return 0;
        }

        int lh = fDiam(root->left,maxi);
        int rh = fDiam(root->right,maxi);

        maxi = max(lh+rh,maxi);
        return 1+ max(lh,rh);
    }
    
    int diameterOfBinaryTree(TreeNode* root) {
        int maxi = 0;
        fDiam(root,maxi);
        return maxi;

        
    }
};